### PR TITLE
(#93) Support simple_app_under_sev to run in simulated SEV mode.

### PR DESCRIPTION
**Pull request by gapisback**
On Linux platforms, we can build and install sev-snp-simulator/ to create a simulated SEV environment.

This commit adds new setup scripts to setup the simulated SEV environment.
And, adds all the required steps to `run_example.sh` to run `simple_app_under_sev` in simulated SEV mode.

The main changes are as follows:

 - To get increased testing coverage, with this commit,
       this simulated-SEV environment is the default mode in
       which 'setup' target to run_example.sh will run for
       simple_app_under_sev.

- Introduced `CI/scripts` dir, with `setup-JSON-schema-validator-for-SEV-apps.sh`
       as the main script to package the pre-requisite setup steps
       to build the Policy Generator executable.

- Added several SEV-related tests to build.yml, including
       build-and-test of simple_app_under_sev in simulated SEV mode.

- Updated SEV- and simulated-SEV platform details in various docs
- Improved messages of test being run via run_example.sh script. This is to better track the voluminous messages generated in CI-runs.

- Consolidated some duplicate *.mak files
- Updated some notes / instructions to md-file syntax.

 Following should execute the simple-app in simulated-SEV env:
```
    $       ./run_example.sh simple_app_under_sev setup
    $ sudo ./run_example.sh simple_app_under_sev run_test
```

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -16,19 +16,35 @@ jobs:
     - name: dependencies
       run: |
         sudo apt update -y
-        sudo apt install -y libgtest-dev libgflags-dev openssl libssl-dev protobuf-compiler protoc-gen-go golang-go
+        sudo apt install -y libgtest-dev libgflags-dev openssl libssl-dev protobuf-compiler protoc-gen-go golang-go cmake
```
Comment (gapisback) [inline .github/workflows/build.yml:19]: Include `cmake` as one of the build steps for JSON-related packages needs it.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -16,19 +16,35 @@ jobs:
...
         make -f certifier_tests.mak
         ./certifier_tests.exe
-        #! ---------------------------------------------------------------------
-        cd ../sample_apps
+        popd
+
+        echo " "
+        echo "******************************************************************"
+        echo "* Exercise run_example with --help, --list arguments ..."
+        echo "******************************************************************"
```
Comment (gapisback) [inline .github/workflows/build.yml:43]: Added several such banner messages, so we can see what's running in CI-outputs.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -74,14 +95,90 @@ jobs:
...
+        echo "* Test: Execute script to compile, build and run simple_app."
+        echo "******************************************************************"
+        echo " "
         ./cleanup.sh
         ./run_example.sh simple_app
         ./cleanup.sh
 
-        #! ---------------------------------------------------------------------
-        #! Few other miscellaneous checks
-        cd ../certifier_service/oelib
```
Comment (gapisback) [inline .github/workflows/build.yml:85]: Moved to the bottom.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -74,14 +95,90 @@ jobs:
...
+        echo "* Check that Certifier builds with simulated SEV-enabled."
+        echo "******************************************************************"
+        echo " "
+        make -f certifier.mak clean
+        make -f certifier_tests.mak clean
+        ENABLE_SEV=1 make -f certifier.mak
+        popd
+
+        #! Run script that will setup s/w required to build Policy Generator for SEV-app
+        ./CI/scripts/setup-JSON-schema-validator-for-SEV-apps.sh
```
Comment (gapisback) [inline .github/workflows/build.yml:162]: Added a new `CI/scripts` dir with one setup script. This packages all the steps required.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -74,14 +95,90 @@ jobs:
...
+
+        echo " "
+        echo "******************************************************************"
+        echo "* Run simple_app_under_sev in simulated-SEV environment."
+        echo "******************************************************************"
+        echo " "
+        pushd ./sample_apps
+
+        ./run_example.sh simple_app_under_sev setup
+        sudo ./run_example.sh simple_app_under_sev run_test
```
Comment (gapisback) [inline .github/workflows/build.yml:172]: The main new value-addition test with this work.

File: INSTALL.md [hunk 6]
```diff
@@ -93,12 +110,15 @@ $ make -f policy_utilities.mak clean
...
-policies, you need to build it in utilities. The Policy Generator requires the
-json-schema-validator library which in turn depends on the JSON for Modern C++
-library.
+policies, you need to build the Generator in the `utilities/` directory.
+The Policy Generator requires the json-schema-validator library which in turn
+depends on the JSON for Modern C++ library.
 
-Follow these instructions to have them built and installed on your system:
+Follow these instructions to have these software packages built and installed
+on your system:
```
Comment (gapisback) [inline INSTALL.md:121]: Note to myself: Xref the new script in `CI/scripts`, as it packages all these steps in one script.

File: sample_apps/run_example.sh [hunk 7]
```diff
@@ -27,8 +27,11 @@ SimpleServer_PID=0
 Valid_app=0         # While looking thru SampleApps[] for supported app
 Found_step=0        # While looking thru Steps[] for user-specified fn-name
 DryRun=0            # --dry-run will turn this ON; off by default
+DryRunTag=" "
 DryRun_msg=""
 
+RunTag=""           # Short description of which app is being tested
```
Comment (gapisback) [inline sample_apps/run_example.sh:33]: Added for improved message outputs.

File: sample_apps/run_example.sh [hunk 8]
```diff
@@ -51,7 +54,17 @@ Simple_App_under_gramine_MbedTLS_dir="sample_apps/simple_app_under_gramine/${Mbe
 
 # Work-instructions for SEV-app need following files pre-populated by the user.
 #
-Platform_data_files_SEV="sev_policy.json ark_cert.der ask_cert.der vcek_cert.der"
+Platform_cert_files_SEV="ark_cert.der ask_cert.der vcek_cert.der"
+Platform_data_files_SEV="sev_policy.json ${Platform_cert_files_SEV}"
+
+# Normal usage (esp for testing) will be on simulated SEV-SNP environment
+# Unknown env; will be determined later on.
+CC_SIMULATED_SEV="${CC_SIMULATED_SEV:- -1}"
```
Comment (gapisback) [inline sample_apps/run_example.sh:62]: For debugging individual steps, you can now run:

```CC_SIMULATED_SEV=1 ./run_example.sh simple_app_under_sev <step-name>```

File: sample_apps/run_example.sh [hunk 9]
```diff
@@ -152,13 +165,22 @@ function usage_OE() {
...
+    echo "  To run and test the example program : sudo ./${Me} simple_app_under_sev run_test"
+    echo " "
+    echo "-- On SEV-SNP enabled hardware platform:"
+    echo "   Pre-populate simple_app_under_sev/platform_data/ with these required files:"
+    echo "   ${Platform_data_files_SEV}"
+    echo " "
+    echo "  To setup the example program        :      ./${Me} simple_app_under_sev setup_with_auto_policy_generation_for_SEV"
+    echo "  To run and test the example program : sudo ./${Me} simple_app_under_sev run_test"
+    echo " "
+    echo "  To cleanup after a run              : sudo ./cleanup.sh"
```
Comment (gapisback) [inline sample_apps/run_example.sh:183]: Rearranged help messages. Run `run_example.sh --help simple_app_under_sev` to see how these messages look like.

File: sample_apps/run_example.sh [hunk 10]
```diff
@@ -392,20 +418,28 @@ function run_cmd() {
...
+    local array_len=0
+    array_len=$((${#args_array[@]}))
+
+    local echo_sudo=""
+    local run_sudo=0
+    if [ "${RunSudo}" -eq 1 ] && [ "${array_len}" -gt "1" ]; then
+        echo_sudo="sudo "
+        run_sudo=1
+    fi
+
```
Comment (gapisback) [inline sample_apps/run_example.sh:438]: You can safely ignore these diffs. These are solely for improved messages indicating what's being run in normal and in `--dry-run` mode.

File: sample_apps/simple_app_under_sev/instructions.md [hunk 11]
```diff
@@ -99,16 +140,16 @@ $CERTIFIER_PROTOTYPE/utilities/simulated_sev_key_generation.exe            \
...
-mv sev_ask_cert.der ask_cert.der
-mv sev_vcek_cert.der vcek_cert.der
+# Save existing cert.der files as .old files
+$ for der in ark_cert.der ask_cert.der vcek_cert.der; do cp -p ${der} ${der}.old; done
+
+# cp over newly created sev*.der files as cert.der files
+$ for der in ark_cert.der ask_cert.der vcek_cert.der; do cp -p sev_${der} ${der}; done
 ```
 
+# RESOLVE: This comment seems wrong ... fix it.
```
Comment (gapisback) [inline sample_apps/simple_app_under_sev/instructions.md:152]: @yelvmw -- Let's discuss and fix this text. Suggest what I need to do to reconcile the docs with what's reality.

File: src/sev_tests.cc [hunk 12]
```diff
@@ -19,6 +19,9 @@
 
 #include "attestation.h"
 
+using namespace certifier::framework;
+using namespace certifier::utilities;
```
Comment (gapisback) [inline src/sev_tests.cc:23]: Otherwise, this file would not compile.

**Comment by gapisback** [comment]
Good morning, @yelvmw -- just checking if you have any comments / feedback on this change set. Would be nice to clear this review, so I can build dependent docs-updates work. Thanks!

**Review by yelvmw — APPROVED** [review]
LGTM. Thanks!

File: INSTALL.md (hunk 6 above)
Comment (gapisback) [reply INSTALL.md:121]: Done. Updated this md to xref the CI-script.

**Pull request merged**